### ISSUE TITLE
[Question] Why do subdecks always generate much smaller parameters than the mother deck?

### ISSUE THREAD

**Issue by JeffersonCorreia32a** [open-spaced-repetition/fsrs4anki#322]
**Research**

*Enter an [x] character to confirm the points below:*

- [x ] I have checked the [FAQ](https://github.com/open-spaced-repetition/fsrs4anki/wiki/FAQ) and could not find an answer to my question
- [x ] I have read the [wiki](https://github.com/open-spaced-repetition/fsrs4anki/wiki) and still felt confused
- [ x] I have searched for similar existing questions here

**Question**

I make this topic to be discussed because I haven't seen anything talking about it here, you who are more in-depth if you can detail my doubt it would be wonderful. The question is the following, I have noticed that depending on the hierarchy or number of cards in the optimized deck, the higher this hierarchy or the total number of cards in that optimized deck, the greater the "w" parameter is, this makes me think about the logic of having larger parameters for the amount of cards inside the deck, for example:

"w" from the mother of all deck: [1.2896, 1.2765, 4.8799, -0.4283, -0.6138, 0.1059, 1.5208, -0.15, 0.9193, 2.078, -0.1203, 0.2652, 1.1044],

"w" from subdeck01: [1.1787, 1.1787, 4.9467, -0.4823, -0.5179, 0.2157, 1.4533, -0.15, 0.8532, 2.0532, -0.1468, 0.2532, 1.0533] ,

"w" from subdeck02: [1.2106, 1.2104, 4.9088, -0.4137, -0.5504, 0.2185, 1.4901, -0.15, 0.89, 2.0697, -0.1306, 0.2863, 1.0546]

"w" from subdeck03: [0.9398, 1.1789, 4.9471, -0.5178, -0.5352, 0.1472, 1.4531, -0.15, 0.853, 2.0186, -0.1774, 0.1824, 1.0531]

What is the logic of the mother deck having parameters much larger than the parameter of its subdeck with a larger parameter? what is the logic of the mother deck having significantly larger parameters than your subdeck02 which is your subdeck with the largest parameter of the 3 subdecks?

I would like to understand this logic, as this is making me doubt whether the optimization algorithm is really coherent...

**Comment by L-M-Sherlock** [comment]
Because these parameters are optimized from the same set of initial parameters. If the number of reviews in your deck is small, the parameters will deviate less from the initial parameters. The number of reviews in the parent deck is the largest.

**Comment by JeffersonCorreia32a** [comment]
@L-M-Sherlock It would be interesting to see the possibility of this algorithm working differently. Such a form would be that it does not start from the parameters of w [1, 1, 5, -0.5, -0.5, 0.2, 1.4, -0.12, 0.8, 2, -0.2, 0.2, 1]. I have this opinion because, firstly, I see that we don't know for sure the amount of cards or revlogs needed to be able to stabilize this optimization. Therefore, let's just assume that this amount is 2000 cards and their revlogs, in this case we will consider 2 examples, 1st example a person who with 2000 cards makes a single optimization of the complete battalion, this person will have an optimization with values ​​greater than a person who, for example, has 20 subdecks with 100 cards each, totaling the same 2000, if that person performs optimization on each deck of 100 cards because he wants a specific optimization for each subdeck, he will have very different parameters than the person who performed only 1 optimization of the full 2000 deck. This, in my opinion, ends up being an inconsistency, I think there should be a way for the model to calculate these optimization parameters without being tied to the number of revlogs considered and without being tied to the number of cards. Even if that amount of revlogs and cards is useful to further refine the "w" parameters, it cannot be so preponderant as to prevent the person from making a parameterization so far from what he should, just because there are few cards in a given deck.
Even because the ideal, as they put in the other topic [Feature Request] Automatic Generation of Multiple Decks' Parameters #314, is that the optimization in the future reaches a way where we automatically manage to optimize each individual deck, since it can occur in having a deck that has cards with different styles and difficulties, such as an exact sciences deck and a legislation deck, which will generate cards that are very different from each other. I would like you to analyze the ideas carefully as this can lead to a very important improvement in the algorithm.

**Comment by L-M-Sherlock** [comment]
I think this problem can be solved by fixing the number of iteration in the training. For example, set the total iteration at 100,000. If we have 2000 cards with 10,000 review logs, the number of epoch is 100,000/10,000 = 10. If we have 100 cards with 500 review logs, the number of epoch is 100,000/500 = 200.

**Comment by Expertium** [comment]
@L-M-Sherlock But that doesn't work if someone has more than 100,00 reviews, and I know I do.

**Comment by JeffersonCorreia32a** [comment]
@L-M-Sherlock, in a hypothetical situation, if I have a deck of 10 cards with a review log history exhibiting characteristics such as retention, difficulty, etc., exactly the same as a deck of 1000 cards where the review log history also possesses the same characteristics, theoretically they should have the same parameter "w" as long as their characteristics remain the same. In other words, in my opinion, the optimizer's result for "w" should be directly linked to the average characteristics of the review logs, regardless of the quantity of logs. The idea would be for the formula to calculate the same value for the "w" parameters whether you have 10 cards or 1000 cards, provided their average characteristics are exactly the same.

Currently, in my case, what happens is that we start with a value for "w" that tends to increase as more cards and logs are created. This prevents me from wanting to optimize decks that I consider to have different styles, as my "w" value would decrease significantly if I were to do so. But if we consider that I am creating and studying cards with the same characteristics within a specific deck, and I remain the same person with the same memory, why should there only be a tendency for "w" to increase over time?

I believe it should start with a value close to my style of review logs, and as I add more cards and have more logs, or if I change my style in a way that also affects the logs, it should gradually improve to match this average style.

Note: I apologize if anything was repetitive. Yesterday, I was sending the message from my smartphone, so please forgive any repetition or errors.

**Comment by Expertium** [comment]
> and as I add more cards and accumulate more review logs, or if I change my style in a way that also affects the review logs, it would gradually refine to align with this average style.

That's what happens. If you are trying to say that the average magnitude of parameters is only getting higher as the number of reviews increases, then no. It does happen to _some_ degree, but not nearly as much as you think. In other words, if you had an infinite amount of reviews, the optimal parameters wouldn't be infinitely large.

**Comment by JeffersonCorreia32a** [comment]
I might not have made myself clear, but the idea is to try to make the "w" value, from the beginning and the initial revlogs, as close as possible to the value it will reach when there are 20,000 revlogs, for example.

In my ideal view, if I have a total of only 10 revlogs, and these 10 revlogs have the same characteristics as the next 10,000, the initial "w" value (when there were 10 revlogs) should be the same or similar to the final "w" value (after 10,000 more revlogs with the same characteristics), rather than starting small and ending up large as it currently happens.

I conducted tests with my decks (topics) and subdecks (subtopics), optimizing them and comparing the "w" values, and I noticed that while my deck reached a "w" value of [1.29...], the subdecks, despite being part of the deck and having the same card characteristics, ended up with lower "w" values of [1.17...]. None of the subdecks were able to achieve the same "w" value as the deck; all subdecks had lower "w" values. This suggests that the quantity of cards had a greater influence than their characteristics. I suggest conducting similar tests with your decks and subdecks to verify if what I'm saying is true.

Now, the question arises: Why does a subdeck with similar cards but a smaller quantity have to have a lower "w" value when it should have the same or at least a similar "w" value (considering that their card characteristics are identical since we are talking about the same topic)? 



Note.:Thank you in advance for your patience with my questions, and I hope that in some way, it can be productive for the developers to consider ways to improve the algorithm.

**Comment by Expertium** [comment]
Post all sets of parameters, please. Both for the main deck and for the subdeck (or multiple subdecks).

**Comment by JeffersonCorreia32a** [comment]
@Expertium  Here are the results of the tests. I admit that although there are still differences, this time they were much smaller compared to other tests I conducted. However, even in this test, although the differences are less significant, they still exist.

What also intrigues me is that the values of the deck "0PROJETO CONCURSO 20mil+::DIREITO ADM." should approximately be an average of the values of its subdecks. However, all the subdecks have lower values.

Deck: 0PROJETO CONCURSO 20mil+::DIREITO ADM.
Card Quantity: 620
"w" values: [1.2106, 1.2104, 4.9043, -0.4041, -0.5319, 0.2263, 1.4943, -0.15, 0.8942, 2.0701, -0.1296, 0.2859, 1.0531]

Deck: 0PROJETO CONCURSO 20mil+::DIREITO ADM.::A00.1-REGIME JURÍDICO ADMINISTRATIVO
Card Quantity: 57
"w" values: [1.1782, 1.089, 4.9468, -0.449, -0.5, 0.2527, 1.4532, -0.15, 0.8531, 2.0533, -0.1467, 0.2533, 1.0532]

Deck: 0PROJETO CONCURSO 20mil+::DIREITO ADM.::A01.1-NOÇÕES INICIAIS DE DIREITO ADM
Card Quantity: 83
"w" values: [1.1776, 1.0717, 4.947, -0.4471, -0.5522, 0.1821, 1.4531, -0.15, 0.8531, 1.9829, -0.2169, 0.185, 0.9822]

Deck: 0PROJETO CONCURSO 20mil+::DIREITO ADM.::A05 PODERES DA ADMINISTRAÇÃO PÚBLICA. PODERES E DEVERES DOS ADMINISTRADORES PÚBLICOS. USO E ABUSO DE PODER
Card Quantity: 130
"w" values: [0.9966, 1.048, 4.9468, -0.4478, -0.4562, 0.2514, 1.4533, -0.15, 0.8532, 2.0388, -0.1568, 0.2531, 1.0528]

Deck: 0PROJETO CONCURSO 20mil+::DIREITO ADM.::A06 ATOS ADMINISTRATIVOS
Card Quantity: 75
"w" values: [1.1775, 1.0461, 4.9467, -0.4467, -0.5179, 0.2531, 1.4534, -0.15, 0.8533, 2.0533, -0.1467, 0.2533, 1.0533]

Deck: 0PROJETO CONCURSO 20mil+::DIREITO ADM.::A12-LEI DE IMPROBIDADE ADMINISTRATIVA
Card Quantity: 275
"w" values: [1.1782, 1.1782, 4.9469, -0.447, -0.5207, 0.212, 1.4532, -0.15, 0.8531, 2.0187, -0.1808, 0.2532, 1.0171]

**Comment by user1823** [comment]
> I might not have made myself clear, but the idea is to try to make the "w" value, from the beginning and the initial revlogs, as close as possible to the value it will reach when there are 20,000 revlogs, for example.

I support the idea. But, the problem is in the technical aspect. The optimizer can't cough up the correct set of "w" values with just the initial few reviews, especially considering that there are 13 different parameters (w) to optimize.

Let me briefly describe the mechanism of optimization to make things clear.
- First, there is a pre-defined set of parameters (which is same for all users).
- Then, with each review processed, the optimizer tries to adjust the parameters in such a way that the loss (which in very simple terms is a measure of the difference between the retention predicted by the model and the actual retention) is minimized.

Now, the amount by which each iteration (each review processed) can affect the value of the parameters is controlled by a value called learning rate. However, using large learning rates can lead to some problems.

So, this means that the optimizer can only slightly adjust the values of the parameters with each iteration. This would imply that there is a minimum number of reviews that the optimizer requires to produce a reasonably good set of parameters. However, note that this minimum number is not well-defined and can vary among different collections.

With that being said, it doesn't mean that we can't make the parameters, from the beginning and the initial revlogs, as close as possible to the values they will eventually reach. We just need to find the appropriate "hack".

One thing I can think of this is the following. First, select few reviews randomly and then use a high learning rate for those reviews to generate an initial set of parameters. Then, do the usual optimization with all the reviews (including ones that were selected before) and with the normal learning rate.
@L-M-Sherlock, would this approach work?

**Comment by user1823** [comment]
Another possible solution:

If n_reviews × n_epoch < 100,000, increase the number of epochs so that the total number of iterations becomes 100,000.

Here,
- n_reviews = total number of reviews in the collection
- n_epoch = default value of n_epoch

This is the same as Sherlock's idea above, but with the added condition "If n_reviews × n_epoch < 100,000". This should also solve Expertium's concern that Sherlock's idea wouldn't work if someone has more than 100,000 reviews.

**Comment by L-M-Sherlock** [comment]
FSRS used [CosineAnnealingLR](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html) to decay the learning rate.

https://github.com/open-spaced-repetition/fsrs4anki/blob/aab50e17c84b6fb82d425879055c1934010aa935/package/fsrs4anki_optimizer/fsrs4anki_optimizer.py#L173

**Comment by JeffersonCorreia32a** [comment]
Thank you, @L-M-Sherlock ! I appreciate it!

I conducted an extreme test to see how much the learning rate was reduced with this change, as I only understand what's happening by conducting practical tests since I don't know programming.

Well, the test was as follows: I have the following deck: Deck: 0PROJETO CONCURSO 20mil+::DIREITO ADM.::A12-LEI DE IMPROBIDADE ADMINISTRATIVA
Number of Cards: 275
"w" values: [1.1782, 1.1782, 4.9469, -0.447, -0.5207, 0.212, 1.4532, -0.15, 0.8531, 2.0187, -0.1808, 0.2532, 1.0171]

From this deck with 275 cards, I removed 14 cards and created a new deck with them. I then optimized this new deck with the transferred 14 cards. The result, in my opinion, was very good because with only 14 cards in the deck, the "w" values were as follows: [1.148, 1.0894, 5.0267, -0.5267, -0.4733, 0.1733, 1.3733, -0.175, 0.7733, 2.0267, -0.1733, 0.2267, 1.0267]

In reality, I don't even know what each parameter of the "w" means, but evaluating the numbers as a layperson, they are not very different from each other. Therefore, I think the result was satisfactory for such an extreme test. In my understanding, this means that if I were to take more cards, the already decent result would get closer and closer. Now I will take another 36 random cards from the original deck of 275 and add them to the 14 cards to make a total of 50 cards, then redo the optimization.

Result obtained with the 50 cards taken from the original 275: "w" values: [1.1784, 1.1785, 5.0527, -0.5178, -0.4822, 0.2173, 1.3788, -0.2533, 0.7784, 1.9874, -0.2042, 0.1826, 0.9835]

I am satisfied with these results. With just 50 cards, the "w" values are very similar. As I mentioned before, I don't know how these differences actually affect scheduling, but from my layperson's perspective, the "w" values became very similar with 50 cards taken from the original 275.

In my layperson's understanding, and please correct me if I'm wrong, the results of my test were excellent!

**Comment by user1823** [comment]
> Thank you, @L-M-Sherlock ! I appreciate it!
> 
> I conducted an extreme test to see how much the learning rate was improved with this change,

But, no such change was implemented recently.

**Comment by Expertium** [comment]
Yes, I checked older files and `CosineAnnealingLR` has been used for at least a couple of months.

**Comment by L-M-Sherlock** [comment]
> If n_reviews × n_epoch < 100,000, increase the number of epochs so that the total number of iterations becomes 100,000.

I'm afraid it would cause overfitting when the n_reviews is very low.

**Comment by JeffersonCorreia32a** [comment]
Acredito que fiz confusão, isso porque conforme foi conversado no tópico "[Feature Request] Automatic Generation of Multiple Decks' Parameters #314", eu fiz algumas otimizações com data bem antiga, sendo que eu tinha reformulado meus decks e provavelmente a otimização estava captando esses revlogs antigos. Após fazer como @user1823 me indicou, utilizar os decks com arquivo .apkg ao invés do .colpkg e colocando o revlog_start_date = "2023-01-04". Após corrigir esse problema, e de acordo os testes que fiz acima, senti os resultados bem mais satisfatórios do que anteriormente. Se algo puder ainda melhorar seria ótimo claro, mas no meu entender está bom. Acho que seria importante ver a questão inicial abordada lá no tópico "[Feature Request] Automatic Generation of Multiple Decks' Parameters #314".

**Comment by L-M-Sherlock** [comment]
> If n_reviews × n_epoch < 100,000, increase the number of epochs so that the total number of iterations becomes 100,000.

I think we can run the benchmark to test this idea now.

**Comment by user1823** [comment]
I think that we should solve this first: https://github.com/open-spaced-repetition/rs-fsrs/issues/10

**Comment by user1823** [comment]
In https://github.com/open-spaced-repetition/fsrs-rs/issues/117, when I said that the other weights will remain largely unchanged, I considered only the current method of optimization.

However, when we are planning to increase the number of epochs so that the total number of iterations becomes 100k, over-fitting can occur if sufficient reviews are not available.

So, we need to rethink about that question (of using more limitations) keeping in mind the proposed optimization method.

**Comment by L-M-Sherlock** [comment]
I prefer the current method.
